src/spec.rs

```diff
@@ -0,0 +1,104 @@
+    pallas::Base::from_bytes_wide(hash.as_array())
+}
+
+/// $\mathsf{ToScalar}^\mathsf{Orchard}(x) := LEOS2IP_{\ell_\mathsf{PRFexpand}}(x) (mod r_P)$
+///
+/// Defined in [Zcash Protocol Spec § 4.2.3: Orchard Key Components][orchardkeycomponents].
+///
+/// [orchardkeycomponents]: https://zips.z.cash/protocol/nu5.pdf#orchardkeycomponents
+pub(crate) fn to_scalar(hash: Hash) -> pallas::Scalar {
+    pallas::Scalar::from_bytes_wide(hash.as_array())
+}
+
+/// Defined in [Zcash Protocol Spec § 4.2.3: Orchard Key Components][orchardkeycomponents].
+///
+/// [orchardkeycomponents]: https://zips.z.cash/protocol/nu5.pdf#orchardkeycomponents
+pub(crate) fn commit_ivk(
+    ak: &pallas::Base,
+    nk: &pallas::Base,
+    rivk: &pallas::Scalar,
+) -> pallas::Scalar {
+    let ivk = sinsemilla::short_commit(
+        "z.cash:Orchard-CommitIvk",
+        iter::empty()
+            .chain(ak.to_le_bits().iter().by_val().take(L_ORCHARD_SCALAR))
+            .chain(nk.to_le_bits().iter().by_val().take(L_ORCHARD_SCALAR)),
+        rivk,
+    );
+
+    // Convert from pallas::Base to pallas::Scalar. This requires no modular reduction
+    // because Pallas' base field is smaller than its scalar field.
+    pallas::Scalar::from_repr(ivk.to_repr()).unwrap()
+}
+
+/// Defined in [Zcash Protocol Spec § 5.4.1.6: DiversifyHash^Sapling and DiversifyHash^Orchard Hash Functions][concretediversifyhash].
+///
+/// [concretediversifyhash]: https://zips.z.cash/protocol/nu5.pdf#concretediversifyhash
+pub(crate) fn diversify_hash(d: &[u8; 11]) -> pallas::Point {
+    pallas::Point::hash_to_curve("z.cash:Orchard-gd")(d)
+}
+
+/// $PRF^\mathsf{expand}(sk, t) := BLAKE2b-512("Zcash_ExpandSeed", sk || t)$
+///
+/// Defined in [Zcash Protocol Spec § 5.4.2: Pseudo Random Functions][concreteprfs].
+///
+/// [concreteprfs]: https://zips.z.cash/protocol/orchard.pdf#concreteprfs
+pub(crate) fn prf_expand(sk: &[u8], t: &[u8]) -> Hash {
+    prf_expand_vec(sk, &[t])
+}
+
+pub(crate) fn prf_expand_vec(sk: &[u8], ts: &[&[u8]]) -> Hash {
+    let mut h = Params::new()
+        .hash_length(64)
+        .personal(PRF_EXPAND_PERSONALIZATION)
+        .to_state();
+    h.update(sk);
+    for t in ts {
+        h.update(t);
+    }
+    h.finalize()
+}
+
+/// Defined in [Zcash Protocol Spec § 5.4.4.5: Orchard Key Agreement][concreteorchardkeyagreement].
+///
+/// [concreteorchardkeyagreement]: https://zips.z.cash/protocol/nu5.pdf#concreteorchardkeyagreement
+pub(crate) fn ka_orchard(sk: &pallas::Scalar, b: &pallas::Point) -> pallas::Point {
+    b * sk
+}
+
+/// Hash extractor for Pallas.
+///
+/// Defined in [Zcash Protocol Spec § 5.4.8.7: Hash Extractor for Pallas][concreteextractorpallas].
+///
+/// [concreteextractorpallas]: https://zips.z.cash/protocol/nu5.pdf#concreteextractorpallas
+pub(crate) fn extract_p(point: &pallas::Point) -> pallas::Base {
+    // TODO: Should we return the actual bits in a Vec, or allow the caller to use
+    // PrimeField::to_le_bits on the returned pallas::Base?
```

Comment:
Either works.

Reply:
The way it is now is closer to the spec. (It changed from returning a bit sequence to a ℙ<sub>x</sub> when I added the ℙ<sub>x</sub> type.)